### DISCUSSION TITLE
Ignoring bot using invalid secret: Corners, version: 1.0

### DISCUSSION THREAD

**Discussion by electricSoda** [robocode-dev/tank-royale#95] [Q&A] [answered]
Hi,

I've been trying to figure out how to get the Websocket functionality to work locally (between 2 computers). I'm using Java and I'm using the Java sample bots.  My issue is when I start up a server on the GUI, then go to the terminal, and go to a sample bot's directory (e.g. Corners), and then run the command, `java -cp ../lib/* Corners.java`, the bot will connect and then disconnect. The error from the server logs gives: `INFO Ignoring bot using invalid secret: Corners, version: 1.0`.

I see in the `server.properties` file generated a property called "bots-secrets". Does this have anything to do with this error?

Any help is appreciated
Thanks!

**Comment by flemming-n-larsen** [comment, marked as answer]
@electricSoda 
I realise that this information is missing in the documentation. Sorry for that. I will make sure that this information is included in the documentation soon so everybody knows what to do.

You are right. Bots can only join the server by providing a bot secret that matches the one created by the server. This is a security constraint so bots cannot join a server without "permission", which is provided through the bot secret.

The easiest way to provide the bot secret is to set the environment variable `BOT_SECRETS` to the bot-secret value from the server. It must be defined in the terminal/shell running the bot. The booter sets this in the terminal (behind the scene) when booting a bot. So when you need to connect to the server outside the server, you need to set the environment variable yourself.

**Reply by electricSoda** [reply]
Hi, thanks so much for your quick response!

The secrets are working now! I just did `set SERVER_SECRET=<your secret>` and it works! I think it would be `export SERVER_SECRET=<your secret>` if you're on Mac.

However, I'm running into another issue, which is when I change the server websocket to something instead of `ws://localhost:7654`, e.g. `ws://192.168.1.20:3000`, I'm finding that I can't start a battle, even though I am able to connect bots booted up from separate computers. 

To reproduce: 
- Go to the `server.properties` file and add your computer's local IP address converted to a websocket address (e.g. `ws://192.168.1.20:3000`), and select that from Server > Select Server from the GUI
- Start up a server on this new websocket address through the GUI
- On a separate computer (or on the same computer), change the SERVER_URL environment variable (like I did with the SERVER_SECRET) to the same websocket address you did in the `server.properties` file
- Then go in command prompt or terminal and then boot up a bot e.g. `java -cp ../lib/* Corners.java`
- It should show that it's connected on the battle menu thingy (in the Joined Bots (local/remote))
- Add that remote bot for battle
- Boot up another bot from a local directory through the GUI (not from command line)
- Click start battle
- You will see that the pop up for selecting bots disappears like it's going to start a match, but then nothing shows up on the main screen (the battle interface doesn't show up)

Do you know why this is?
(Also do you want me to make this a separate Q&A discussion?)

**Reply by flemming-n-larsen** [reply]
@electricSoda 

Yes, I should like you to create a bug report, as this is a new topic and potential a bug (I haven't tried it out yet).
This way other developers can see the issue as well, and it is easier to keep track of issues in general, and also prioritize what to work on next. 🙂

A Bug report can be created from here:
https://github.com/robocode-dev/tank-royale/issues/new/choose

Just fill in the data from this thread, as I think it is sufficient.

**Comment by flemming-n-larsen** [comment]
I realised that the javadocs has been hiding protected members. Hence, I fixed that so  the available constructors for the Bot can be seen:

[Bot constructors](https://robocode-dev.github.io/tank-royale/api/java/dev/robocode/tankroyale/botapi/Bot.html)

This constructor takes a server secret. This way it is possible to provide the Bot with a server secret programmatically:
[protected BaseBot​(BotInfo botInfo, URI serverUrl, String serverSecret)](https://robocode-dev.github.io/tank-royale/api/java/dev/robocode/tankroyale/botapi/BaseBot.html#%3Cinit%3E(dev.robocode.tankroyale.botapi.BotInfo,java.net.URI,java.lang.String))
